File: ArchitectureSomeMoreMultiThreading/factory/pinSharpChecker.go
package factory

import (
	"fmt"
	"go.uber.org/atomic"
	"sync"
)

type PinSharpChecker struct {
	// Pins which sharpness is currently being checked.
	pins []*Pin
	// Thread lock.
	lock sync.Mutex
	// Number of total pins, which are still in process of approving / sharpening.
	pinsLeft *atomic.Uint32
	// The resulting vector of pins.
	ans []*Pin
}

// Constructor.
func NewCheck(pinsLeft *atomic.Uint32) *PinSharpChecker {
	checker := new(PinSharpChecker)
	checker.pinsLeft = pinsLeft
	return checker
}

// Thread run.
func (checker *PinSharpChecker) Run(wg *sync.WaitGroup) {
	// Worker has done his job.
	defer wg.Done()
	for checker.pinsLeft.Load() > 0 {
		checker.lock.Lock()

		// If worker has no pins for now.
		if len(checker.pins) == 0 {
			checker.lock.Unlock()
			continue
		}

		if checker.pins[len(checker.pins)-1].sharpness > 0.75 {
			// Good pin.
			checker.lock.Unlock()
			checker.returnPin()
			checker.lock.Lock()
		} else {
			// Bad pin.
			// fmt.Printf("Sharp checker disapproved and threw away a pin with curvature %f and sharpness %f\n",
			//	 checker.pins[len(checker.pins)-1].curvature,
			//	 checker.pins[len(checker.pins)-1].sharpness)
			checker.pinsLeft.Dec()
		}

		// The pin was sent.
		checker.pins = checker.pins[:len(checker.pins)-1]
		checker.lock.Unlock()
	}
}

// Getting a new pin.
func (checker *PinSharpChecker) receivePin(pin *Pin) {
	checker.lock.Lock()
	// Adding a new pin to our current collection.
	checker.pins = append(checker.pins, pin)
	checker.lock.Unlock()
}

// Giving away an approved pin.
func (checker *PinSharpChecker) returnPin() {
	checker.lock.Lock()
	// fmt.Printf("Sharp checker approved a pin with curvature %f and sharpness %f\n",
	//	checker.pins[len(checker.pins)-1].curvature,
	//	checker.pins[len(checker.pins)-1].sharpness)
	checker.pinsLeft.Dec()
	// Updating result.
	checker.ans = append(checker.ans, checker.pins[len(checker.pins)-1])
	checker.lock.Unlock()
}

// Outputting the answer.
func (checker *PinSharpChecker) Output() {
	fmt.Println("\n\nPins, which were successfully sharpened:")
	for i := 0; i < len(checker.ans); i++ {
		fmt.Printf("%d). %s\n", i, checker.ans[i].ToString())
	}
}

File: ArchitectureHW5/pinSharpChecker.cpp
#include <iostream>
#include "pinSharpChecker.h"

// Constructor.
pinSharpChecker::pinSharpChecker(std::atomic<int> *pins_left, std::vector<pin *> *ans) {
    pins_left_ = pins_left;
    ans_ = ans;
}

// Thread run.
void pinSharpChecker::run() {
    while (true) {
        lock_.lock();
        if (*pins_left_ < 1) {
            lock_.unlock();
            return;
        }

        // Checking if program still needs to work.
        if (*pins_left_ < 1) {
            lock_.unlock();
            return;
        }

        // If no new pins came for now.
        if ((pins_.empty()) || (pins_.back() == nullptr)) {
            lock_.unlock();
            continue;
        }

        if (pins_.back()->sharpness > 0.75) {
            lock_.unlock();
            // If we got a good pin, we send it to the resulting vector.
            returnPin();
        } else {
            // Otherwise, we throw it away.
//            std::cout << "Sharp checker disapproved and threw away a pin with curvature " +
//                      std::to_string(pins_.back()->curvature) + " and sharpness " +
//                      std::to_string(pins_.back()->sharpness) + '\n';
            --(*pins_left_);
            lock_.unlock();
        }

        lock_.lock();
        pins_.pop_back();
        lock_.unlock();
    }
}

// Getting a new pin.
void pinSharpChecker::receivePin(pin *pn) {
    lock_.lock();

    // Checking if program still needs to work.
    if (*pins_left_ < 1) {
        lock_.unlock();
        return;
    }

    pins_.push_back(pn);
    lock_.unlock();
}

// Fully approving a pin and putting it in the resulting vector.
void pinSharpChecker::returnPin() {
    lock_.lock();

//    std::cout << "Sharp checker approved a pin with curvature " +
//              std::to_string(pins_.back()->curvature) + " and sharpness " +
//              std::to_string(pins_.back()->sharpness) + '\n';

    --(*pins_left_);
//    std::cout << *pins_left_ << '\n';

    (*ans_).push_back(pins_.back());
    lock_.unlock();
}
